src/lsp/cobol_indent/cobol_indent.ml

```diff
@@ -43,17 +43,16 @@ let apply =
        [String.index_from] can also raise [Invalid_argument] in these
        conditions. *)
     match irs with
+    | [] when start = 0 -> s
     | [] ->
-      if start = 0 then s else (
-        Buffer.add_substring buf s start (String.length s - start);
-        Buffer.contents buf
-      )
+      Buffer.add_substring buf s start (String.length s - start);
+      Buffer.contents buf
     | { lnum; offset_orig; offset_modif } :: irs when line = lnum - 1 ->
       let eol =
         try String.index_from s bol '\n' + 1 with Not_found -> String.length s
       in
       let delta = offset_modif - offset_orig in
-      if delta > 0 || true then (
+      if delta > 0 then (
```

Comment:
Are you sure this is the intended change? I'd got with simply removing the condition and only keeping the `then` branch, rather than trying to optimize the loop out in case `delta ≤ 0` (with a comment "no-op if `delta <= 0`): I personally think that, in general, less lines of code is usually preferable.

Reply:
When `delta < 0` it's not a no-op, we do need to remove the spaces.

Reply:
I mean we do need to remove the characters, we don't know that they are spaces at this point (but the indenter will only remove actual spaces).

Reply:
Ok I see now. The `|| true` made me think the other case was not needed as all the tests seemed to pass). What this actually reveals is that we miss at least one test where the indenter actually need to un-indent.

Please could you add at least one that exercises the second branch? In `test/lsp/lsp_formatting.ml` might be a good place.

Reply:
As I mentioned above we actually don't test this at all. It is only used by the `superbol indent` command, *not* by the LSP (in the LSP the actual edits are made by the editor using `TextEdit`, and in the tests we test the LSP interface with the `Lsp.Text_document.apply_text_document_edits` method — but `Cobol_indent` does not depend on the LSP.

That said — we indeed have no tests where the indenter has to de-indent. Added one.

Reply:
Yes we don't actually test `superbol indent` at all (or the relevant portions of `Cobol_indent`). Then, to finalize this PR, I suggest to proceed as follows:
- move the `apply` function out of `cobol_indent.ml` (where we should typically have a minimal set of module bindings, and if well justified for the library API, module inclusions — the current contents of this file does not follow this rule yet), maybe in a `indent_string.ml` — no need for a module binding in `cobol_indent.ml`, unless it is accessed from outside of the library;
- add a comment around `apply` that it is not tested yet.

Then I think this PR can be merged.

Reply:
Well it is used by `command_indent_file.ml` in `superbol_free_lib`. Although I can easily introduce a wrapper function `indent_range_to_string` or some such instead.

(My personal opinion is that we shouldn't have this code path at all and should just have `cobol_indent` depend on `lsp` but I get that we may not want that)

Reply:
Yes, that'd be quite a large dependency for such a library… so it's debatable.

If the value is used then I'd either bind the new module directly so one can `Cobol_indent.String.apply`, or else just the value like `Cobol_indent.apply_on_string`, or something similar. The idea is to end-up with a clean interface for the library, with function names that are not too redundant as they should typically be called with full qualification.

Reply:
I exposed the `indent_range_str` instead. `apply` is only indented to be used on the result of `indent_range` anyways.

> Yes, that'd be quite a large dependency for such a library… so it's debatable.

I think the point is moot, because realistically there will be no project that depend on `Cobol_indent` and not the rest of Superbol. Anyways, the code exists and works, so it doesn't really matter.

Reply:
> there will be no project that depend on `Cobol_indent` and not the rest of Superbol.

I wouldn't be so sure. One may want a "small" executable that only focuses on indentation (like `ocp-indent`), that some other editor like GNU/Emacs can use in place of the full-blown LSP for "lighter" tasks. For instance, OCaml I use a combination of `merlin` and `ocp-indent`, because I find that combination more customizable.

Anyways, `cobol_indent` is a small library, so I think we can merge at this point.

Reply:
This feels academic since the design of the project is to provide a single super-charged executable with subcommands though; most people would probably just use `superbol indent`.

Reply:
Now that I think about it, the code for effectively indenting a file should just not belong to `cobol_indent`; there, we should just stick to the indentation logic/algorithm. By that approach, `Superbol_free_lib` may just rely on `lsp` to apply the text editions.

(Apologies if there was an `indent_file` function already there; it should have been removed before).